std/range/package.d

```diff
@@ -8606,6 +8606,385 @@ if (isForwardRange!Source && hasLength!Source)
+            }
+
+            /// Ditto
+            void popFront()
+            {
+                assert(!empty, "Attempting to popFront an empty range");
+
+                // source range is longer than index
+                if (_endIndexes.empty && !_source.empty)
+                {
+                    _isEmpty = true;
+                    return;
+                }
+
+                _source.popFrontN(_endIndexes.front()-_offset);
+                _offset = _endIndexes.front();
+                _endIndexes.popFront();
+
+                if (_maxLen)
+                {
+                    _maxLen.decrease();
+                    _isEmpty = _maxLen.empty;
+                }
+
+                // length of the source range is smaller than or equal to the index
+                if (_source.empty)
+                    _isEmpty = true;
+            }
+
+            /// Forward range primitives
+            @property typeof(this) save()
+            {
+                return typeof(this)(_source, _endIndexes, _offset);
+            }
+
+            static if (isInfinite!Source && hasLength!Indexes)
+            /// Length is only provided for infinite sources, when hasLength!Indexes is true
+                @property size_t length()
+                {
+                    return _endIndexes.length + 1;
+                }
+
+            static if (hasSlicing!Source && hasSlicing!Indexes && hasLength!Indexes)
+            {
+                /**
+                Indexing and slicing operations are only provided when both Source and Indexes
+                have slicing and hasLength!Indexes is true.
+                */
+                auto opIndex(size_t index)
+                {
+                    assert(index<=_endIndexes.length, "Chunks index out of bounds");
+
+                    const size_t startIndex = (index == 0 ? 0 : _endIndexes[index-1] - _offset);
+
+                    static if (hasLength!Source)
+                        assert(startIndex<=_source.length, "Chunks index out of bounds");
+
+                    if (index == _endIndexes.length)
+                        return _source[startIndex .. $];
+
+                    static if (hasLength!Source)
+                    {
+                        import std.algorithm.comparison : min;
+                        const size_t endIndex = min(_endIndexes[index] - _offset, source.length);
+                    }
+                    else
+                        const size_t endIndex = _endIndexes[index] - _offset;
+
+                    return _source[startIndex .. endIndex];
+                }
+
+                /// Ditto
+                auto opSlice(size_t lower, size_t upper)
+                {
+                    assert(lower <= upper, "Start index must be lower than end index");
+                    assert(lower <= _endIndexes.length+1 && upper <= _endIndexes.length+1,
+                        "Chunks index out of bounds");
```

Comment:
You should have a `length` member always in this set of circumstances (I think), and check that instead. See trickiness I cited above.